orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.java

```diff
@@ -199,16 +210,36 @@ public ArtifactResolver(
 
   public @Nonnull List<Artifact> getArtifactsForPipelineId(
       @Nonnull String pipelineId, @Nonnull ExecutionCriteria criteria) {
-    Execution execution =
-        executionRepository.retrievePipelinesForPipelineConfigId(pipelineId, criteria)
-            .subscribeOn(Schedulers.io()).toSortedList(startTimeOrId).toBlocking().single().stream()
-            .findFirst()
-            .orElse(null);
+    Execution execution = getExecutionForPipelineId(pipelineId, criteria);
 
     return execution == null ? Collections.emptyList() : getAllArtifacts(execution);
   }
 
-  public void resolveArtifacts(@Nonnull Map<String, Object> pipeline) {
+  public @Nonnull List<Artifact> getArtifactsForPipelineIdWithoutStageRef(
+      @Nonnull String pipelineId, @Nonnull String stageRef, @Nonnull ExecutionCriteria criteria) {
+    List<Artifact> artifacts = getArtifactsForPipelineId(pipelineId, criteria);
```

Comment:
Would it maybe be a bit cleaner (and a bit faster) to just directly call the new function you wrote `getAllArtifacts()` with a filter excluding the stage `it -> !stageRef.equals(it.getRefId())` than to get all artifacts, then all artifacts that match, and subtract the sets?

Reply:
only thing we lose is the log line, but i guess we could omit the use of that.

Reply:
Yeah, I don't think the log here is too important (at least not important enough to justify the extra code)